### PR TITLE
Add parameter validation to the cmac module.

### PR DESCRIPTION
**Pull request by AndrzejKurek**
Summary: Building on #1643, this PR adds parameter validation to the cmac module.

### CODE REVIEW

**Comment by AndrzejKurek** [comment]
@hanno-arm There already is a dedicated test that checks for null parameters of each function, it just does not use the new parameter validation test error checking scheme. As for the first point, I personally prefer not to double checks, but I understand that if you want to have a uniform behaviour of calling the callback if any input error happens, I should add them here too. I'll wait for opinions from others.

**Comment by mpg** [comment]
@AndrzejKurek I think the checks with `CMAC_VALIDATE_RET()` should _replace_ the existing `NULL` checks. See what Hanno did with ChachaPoly in #2275

File: include/mbedtls/cmac.h [hunk 1]
```diff
@@ -96,7 +96,8 @@ int mbedtls_cipher_cmac_starts( mbedtls_cipher_context_t *ctx,
  *                      Can be called repeatedly.
  *
  * \param ctx           The cipher context used for the CMAC operation.
- * \param input         The buffer holding the input data.
+ *                      This must be initialized.
```
[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply include/mbedtls/cmac.h]: I'm not sure if we should duplicate the information that is just above.

File: include/mbedtls/cmac.h [hunk 2]
```diff
@@ -115,7 +116,9 @@ int mbedtls_cipher_cmac_update( mbedtls_cipher_context_t *ctx,
  *                      mbedtls_cipher_cmac_update(), or mbedtls_cipher_free().
  *
  * \param ctx           The cipher context used for the CMAC operation.
+ *                      This must be initialized.
```
[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply include/mbedtls/cmac.h]: I'm not sure if we should duplicate the information that is just above. Come to think of it, I'd even remove it.

**Comment by AndrzejKurek** [comment]
@hanno-arm - I've updated this PR. Please re-review.

File: tests/suites/test_suite_cmac.function [hunk 3]
```diff
@@ -27,71 +27,56 @@ void mbedtls_cmac_null_args(  )
     mbedtls_cipher_init( &ctx );
 
     /* Test NULL cipher info */
-    TEST_ASSERT( mbedtls_cipher_cmac_update( &ctx, test_data, 16 ) ==
-                                         MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+    TEST_INVALID_PARAM( mbedtls_cipher_cmac_update( &ctx, test_data, 16 ) );
```
Comment (k-stachowiak) [inline tests/suites/test_suite_cmac.function]: It appears that the test against the null `cipher_info` validation is only executed for the `..._update()` function and not the others.

File: tests/suites/test_suite_cmac.function [hunk 4]
```diff
@@ -27,71 +27,56 @@ void mbedtls_cmac_null_args(  )
...
 
     cipher_info = mbedtls_cipher_info_from_type( MBEDTLS_CIPHER_AES_128_ECB );
-    TEST_ASSERT( mbedtls_cipher_setup( &ctx, cipher_info ) == 0 );
+    TEST_ASSERT( mbedtls_cipher_setup( &ctx, cipher_info ) == 0 );
```
Comment (k-stachowiak) [inline tests/suites/test_suite_cmac.function]: A trailing white space has crept in here.

File: include/mbedtls/cmac.h [hunk 5]
```diff
@@ -97,7 +98,9 @@ int mbedtls_cipher_cmac_starts( mbedtls_cipher_context_t *ctx,
  *
  * \param ctx           The cipher context used for the CMAC operation.
  *                      This must be initialized.
- * \param input         The buffer holding the input data. This must not be \c NULL.
+ * \param input         The buffer holding the input data. This must be a
+ *                      readable buffer of length \p ilen Bytes. It may be
+ *                      \c NULL if ilen == 0.
```
Comment (hanno-arm) [inline include/mbedtls/cmac.h]: Minor: Use backticks for formatting.

File: include/mbedtls/cmac.h [hunk 6]
```diff
@@ -158,9 +161,12 @@ int mbedtls_cipher_cmac_reset( mbedtls_cipher_context_t *ctx );
  * \param cipher_info   The cipher information. This must not be \c NULL.
  * \param key           The CMAC key. This must not be \c NULL.
  * \param keylen        The length of the CMAC key in bits.
- * \param input         The buffer holding the input data. This must not be \c NULL.
+ * \param input         The buffer holding the input data. This must be a
+ *                      readable buffer of length \p ilen Bytes. It may be
+ *                      \c NULL if ilen == 0.
```
Comment (hanno-arm) [inline include/mbedtls/cmac.h]: Minor: Use backticks for formatting.

File: include/mbedtls/cmac.h [hunk 7]
```diff
@@ -158,9 +161,12 @@ int mbedtls_cipher_cmac_reset( mbedtls_cipher_context_t *ctx );
...
  * \param ilen          The length of the input data.
- * \param output        The buffer for the generic CMAC result. This must not be \c NULL.
+ * \param output        The buffer for the generic CMAC result.
+ *                      This must not be \c NULL.
```
Comment (hanno-arm) [inline include/mbedtls/cmac.h]: Minor: This too should be of the form "This must be a writable buffer ..."

File: library/cmac.c [hunk 8]
```diff
@@ -247,9 +252,9 @@ int mbedtls_cipher_cmac_update( mbedtls_cipher_context_t *ctx,
     int ret = 0;
     size_t n, j, olen, block_size;
 
-    if( ctx == NULL || ctx->cipher_info == NULL || input == NULL ||
-        ctx->cmac_ctx == NULL )
-        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+    MBEDTLS_CMAC_VALIDATE_RET( ctx != NULL );
+    MBEDTLS_CMAC_VALIDATE_RET( ctx->cipher_info != NULL );
+    MBEDTLS_CMAC_VALIDATE_RET( input != NULL );
```
Comment (hanno-arm) [inline library/cmac.c]: It should be `ilen == 0 || input != NULL`.

File: library/cmac.c [hunk 9]
```diff
@@ -395,8 +401,10 @@ int mbedtls_cipher_cmac( const mbedtls_cipher_info_t *cipher_info,
     mbedtls_cipher_context_t ctx;
     int ret;
 
-    if( cipher_info == NULL || key == NULL || input == NULL || output == NULL )
-        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+    MBEDTLS_CMAC_VALIDATE_RET( cipher_info != NULL );
+    MBEDTLS_CMAC_VALIDATE_RET( key != NULL );
+    MBEDTLS_CMAC_VALIDATE_RET( input != NULL );
```
Comment (hanno-arm) [inline library/cmac.c]: It should be `ilen == 0 || input != NULL`.

File: library/cmac.c [hunk 10]
```diff
@@ -432,8 +440,9 @@ int mbedtls_aes_cmac_prf_128( const unsigned char *key, size_t key_length,
     unsigned char zero_key[MBEDTLS_AES_BLOCK_SIZE];
     unsigned char int_key[MBEDTLS_AES_BLOCK_SIZE];
 
-    if( key == NULL || input == NULL || output == NULL )
-        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+	MBEDTLS_CMAC_VALIDATE_RET( key != NULL );
+	MBEDTLS_CMAC_VALIDATE_RET( input != NULL );
```
Comment (hanno-arm) [inline library/cmac.c]: It should be `ilen == 0 || input != NULL`.

File: include/mbedtls/cmac.h [hunk 11]
```diff
@@ -115,7 +119,9 @@ int mbedtls_cipher_cmac_update( mbedtls_cipher_context_t *ctx,
  *                      mbedtls_cipher_cmac_update(), or mbedtls_cipher_free().
  *
  * \param ctx           The cipher context used for the CMAC operation.
+ *                      This must be initialized.
  * \param output        The output buffer for the CMAC checksum result.
```
Comment (hanno-arm) [inline include/mbedtls/cmac.h]: We must document the required size of `output` here. To my understanding, it's the block-size of the underlying cipher.

File: include/mbedtls/cmac.h [hunk 12]
```diff
@@ -151,12 +158,15 @@ int mbedtls_cipher_cmac_reset( mbedtls_cipher_context_t *ctx );
...
- * \param key           The CMAC key.
+ * \param cipher_info   The cipher information. This must not be \c NULL.
+ * \param key           The CMAC key. This must not be \c NULL.
  * \param keylen        The length of the CMAC key in bits.
- * \param input         The buffer holding the input data.
+ * \param input         The buffer holding the input data. This must be a
+ *                      readable buffer of length \p ilen Bytes. It may be
+ *                      \c NULL if ilen == 0.
  * \param ilen          The length of the input data.
  * \param output        The buffer for the generic CMAC result.
```
Comment (hanno-arm) [inline include/mbedtls/cmac.h]: We must document the required size of `output` here. To my understanding, it's the block-size of the underlying cipher.

File: include/mbedtls/cmac.h [hunk 13]
```diff
@@ -176,12 +186,12 @@ int mbedtls_cipher_cmac( const mbedtls_cipher_info_t *cipher_info,
  *                  (AES-CMAC-PRF-128) Algorithm for the Internet Key
  *                  Exchange Protocol (IKE).</em>
  *
- * \param key       The key to use.
+ * \param key       The key to use. This must not be \c NULL.
  * \param key_len   The key length in Bytes.
- * \param input     The buffer holding the input data.
+ * \param input     The buffer holding the input data. This must not be \c NULL.
  * \param in_len    The length of the input data in Bytes.
  * \param output    The buffer holding the generated 16 Bytes of
```
Comment (hanno-arm) [inline include/mbedtls/cmac.h]: Suggestion: This must be a writable buffer of size \c 16 Bytes.

**Comment by sbutcher-arm** [comment]
PR #1643 has been merged, so removing the `needs: preceding PR` label.

File: library/cmac.c [hunk 14]
```diff
@@ -68,6 +68,10 @@
 
 #if !defined(MBEDTLS_CMAC_ALT) || defined(MBEDTLS_SELF_TEST)
 
+#define MBEDTLS_CMAC_VALIDATE_RET(cond) \
```
Comment (sbutcher-arm) [inline library/cmac.c]: These are breaking `check_names.sh` and need to have the `MBEDTLS_` prefix removed.

**Comment by AndrzejKurek** [comment]
I've updated the PR with requested changes, fixed a pre-existing documentation bug (wrong error mentioned in the header file), and fixed & extended the tests. @k-stachowiak, @hanno-arm - please re-review.

**Pull request closed without merging**